src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala

```diff
@@ -181,55 +181,57 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
-                  s"$id: Amount in = $inAmount, out = $outAmount. Allowed new asset = $newAssetId, out = $outAssetId")
-            }
-            .map(_ + totalAssetsAccessCost)
-        case Failure(e) =>
-          // should never be here as far as we've already checked this when we've created the box
-          ModifierValidator.fatal(e.getMessage)
+        extractAssets(boxesToSpend) match {
+          case Success((inAssets, inAssetsNum)) =>
+            lazy val newAssetId = ByteArrayWrapper(inputs.head.boxId)
+            val tokenAccessCost = stateContext.currentParameters.tokenAccessCost
+            val totalAssetsAccessCost = (outAssetsNum + inAssetsNum) * tokenAccessCost +
+              (inAssets.size + outAssets.size) * tokenAccessCost
+            val currentTxCost = validation.result.payload.get
+            val newRemainingCost = remainingCost - currentTxCost - totalAssetsAccessCost
+
+            validation
+              // Check that transaction is not too costly considering all the assets
+              .validate(bsBlockTransactionsCost, newRemainingCost >= 0, s"$id: assets cost")
+              .validateSeq(outAssets) {
+                case (validationState, (outAssetId, outAmount)) =>
+                  val inAmount: Long = inAssets.getOrElse(outAssetId, -1L)
+
+                  // Check that for each asset output amount is no more than input amount,
+                  // with a possible exception for a new asset created by the transaction
+                  validationState.validate(txAssetsPreservation,
+                    inAmount >= outAmount || (outAssetId == newAssetId && outAmount > 0),
+                    s"$id: Amount in = $inAmount, out = $outAmount. Allowed new asset = $newAssetId, out = $outAssetId")
+              }
+              .map(_ + totalAssetsAccessCost)
+          case Failure(e) =>
+            // should never be here as far as we've already checked this when we've created the box
+            ModifierValidator.fatal(e.getMessage)
+        }
       }
-    }
       // Check inputs, the most expensive check usually, so done last.
       .validateSeq(boxesToSpend.zipWithIndex) { case (validation, (box, idx)) =>
+      val currentTxCost = validation.result.payload.get
+      val newRemainingCost = remainingCost - currentTxCost
 
       val input = inputs(idx)
       val proof = input.spendingProof
       val proverExtension = proof.extension
       val transactionContext = TransactionContext(boxesToSpend, dataBoxes, this, idx.toShort)
-      val ctx = new ErgoContext(stateContext, transactionContext, proverExtension)
+      val ctx = new ErgoContext(stateContext, transactionContext, proverExtension, newRemainingCost)
 
       val costTry = verifier.verify(box.ergoTree, ctx, proof, messageToSign)
       costTry.recover { case t => t.printStackTrace() }
 
       lazy val (isCostValid, scriptCost) = costTry.getOrElse((false, 0L))
```

Comment:
Looks like we need to treat failure not as a zero cost computation, but as a maximum cost computation. 
In particular because in the case when the limit has been exceeded, we get an exception in return, and need to add at least that limit to the total cost of the transaction.